third_party/move/move-vm/runtime/src/storage/implementations/unsync_code_storage.rs

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    logging::expect_no_verification_errors,
+    script_hash,
+    storage::{
+        environment::{ambassador_impl_WithRuntimeEnvironment, WithRuntimeEnvironment},
+        implementations::unsync_module_storage::AsUnsyncModuleStorage,
+        module_storage::ambassador_impl_ModuleStorage,
+    },
+    CodeStorage, Module, ModuleStorage, RuntimeEnvironment, Script, UnsyncModuleStorage,
+};
+use ambassador::Delegate;
+use bytes::Bytes;
+use move_binary_format::{
+    access::ScriptAccess, errors::VMResult, file_format::CompiledScript, CompiledModule,
+};
+use move_core_types::{account_address::AccountAddress, identifier::IdentStr, metadata::Metadata};
+use move_vm_types::{code_storage::ModuleBytesStorage, module_linker_error};
+#[cfg(test)]
+use std::collections::BTreeSet;
+use std::{
+    cell::RefCell,
+    collections::{hash_map, HashMap},
+    sync::Arc,
+};
+
+/// Represents an entry in script cache, either deserialized or verified.
+#[derive(Debug)]
+enum ScriptStorageEntry {
+    Deserialized(Arc<CompiledScript>),
+    Verified(Arc<Script>),
+}
+
+/// Code storage that stores both modules and scripts (not thread-safe).
+#[derive(Delegate)]
+#[delegate(
+    WithRuntimeEnvironment,
+    target = "module_storage",
+    where = "M: ModuleStorage"
+)]
+#[delegate(ModuleStorage, target = "module_storage", where = "M: ModuleStorage")]
+pub struct UnsyncCodeStorage<M> {
```

Comment:
Why do we need this and the unsync storage impl in the BlockSTM? Can we use the same unsync storage?

Reply:
Which two are you referring to? We have
- unsync for unit tests, adapter for state view, to handle module publishing 
- block-stm one

Unsync is implemented as module storage, and there is a wrapper which turns it into code storage (so can also handle scripts).